lib/improver/utilities/temporal_interpolation.py

```diff
@@ -0,0 +1,419 @@
+                    ' divide into the interval equally.')
+
+            time_entry = initial_time
+            while True:
+                time_entry = (time_entry +
+                              timedelta(minutes=self.interval_in_minutes))
+                if time_entry >= final_time:
+                    break
+                time_list.append(time_entry)
+
+        return [('time', time_list)]
+
+    @staticmethod
+    def check_cube_coords_dtype(cube, new_cube, dim_coords=None):
+        """
+        Update the data type of the coords within the new_cube
+        to those in the original cube. This can be limited to a list of
+        dimension coordinates and associated auxillary coordinates if required.
+
+        Args:
+        cube (iris.cube.Cube):
+            The original cube that will be used to checked against for the
+            required data type of that coordinate in the new_cube.
+        new_cube (iris.cube.Cube):
+            The cube that must be checked and adjusted using the coordinates
+            from the original cube.
+        dim_coords (list):
+            List of dimension coordinate names. This list will be used to
+            limit the correction of the coordinate data type
+            to these coordinates and associated auxillary coordinates.
+            Default is None.
+
+        Returns:
+        new_cube (iris.cube.Cube):
+            Modified cube with the relevant coordinates datatypes
+            set to the same as on the original cube.
+
+        Raises:
+            CoordinateNotFoundError : If original cube does not have the
+                coordinate in dim_coords.
+            CoordinateNotFoundError : If coordinate in original cube is
+                not in new_cube.
+        """
+        if dim_coords is None:
+            coord_list = [coord.name() for coord in cube.coords()]
+        else:
+            coord_list = []
+            for coord_name in dim_coords:
+                try:
+                    req_dim = cube.coord_dims(coord_name)
+                    associated_coords = [crd.name() for crd in cube.coords()
+                                         if cube.coord_dims(crd) == req_dim
+                                         and crd.name() is not coord_name]
+                    coord_list.append(coord_name)
+                    for val in associated_coords:
+                        coord_list.append(val)
+                except CoordinateNotFoundError:
+                    msg = ('Original cube does not have the coordinate in'
+                           ' dim_coords {}'.format(coord_name))
+                    raise CoordinateNotFoundError(msg)
+        for coord_name in coord_list:
+            dtype_orig = cube.coord(coord_name).points.dtype
+            try:
+                new_coord = new_cube.coord(coord_name)
+                check_coord_datatypes(new_coord, dtype_orig, fix=True,
+                                      rounding=True)
```

Comment:
As commented above, I think the function `check_coord_datatypes` is a bit dangerous and needs rethinking.

Reply:
I'll leave Gavin to comment on the generic function check_coord_datatypes. The data we are using has time in seconds so has no issue with creating data on the half hour.  If the start and end cubes had half hourly data and  units in hours that wouldn't be possible.  So the only real danger is if  we request an interpolated time on the half hour. I tried creating a start and end cube with hourly time coord and requesting a half hour interpolated time however I ran into a small problem, iris_time_to_datetime converts the coord in the original cube to float64 when it runs on time in hourly form and so it never tries to revert the data to int64.